plat/hisilicon/hikey960/hikey960_pm.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2017, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <arch_helpers.h>
+#include <assert.h>
+#include <cci.h>
+#include <console.h>
+#include <debug.h>
+#include <gicv2.h>
+#include <hi3660.h>
+#include <hi3660_crg.h>
+#include <mmio.h>
+#include <psci.h>
+#include "drivers/pwrc/hisi_pwrc.h"
+
+#include "hikey960_def.h"
+
+#define CORE_PWR_STATE(state) \
+	((state)->pwr_domain_state[MPIDR_AFFLVL0])
+#define CLUSTER_PWR_STATE(state) \
+	((state)->pwr_domain_state[MPIDR_AFFLVL1])
+#define SYSTEM_PWR_STATE(state) \
+	((state)->pwr_domain_state[PLAT_MAX_PWR_LVL])
+
+static uintptr_t hikey960_sec_entrypoint;
+
+static void hikey960_pwr_domain_standby(plat_local_state_t cpu_state)
+{
+	unsigned long scr;
+	unsigned int val;
+
+	assert(cpu_state == PLAT_MAX_RET_STATE);
+
+	scr = read_scr_el3();
+
+	/* Enable Physical IRQ and FIQ to wake the CPU*/
+	write_scr_el3(scr | SCR_IRQ_BIT | SCR_FIQ_BIT);
+
+	__asm__ volatile("mrs %0, s3_1_c15_c2_1\n\t"
+			 "bic %0, %0, #0x7\n\t"
+			 "orr %0, %0, #0x2\n\t"
+			 "msr s3_1_c15_c2_1, %0\n\t"
+			 "isb\n\t"
+			 "dsb sy\n\t"
+			 "wfi\n\t"
+			 "mrs %0, s3_1_c15_c2_1\n\t"
+			 "bic %0, %0, #0x7\n\t"
+			 "msr s3_1_c15_c2_1, %0\n\t"
+			 "isb\n\t"
+			 "dsb sy\n\t"
+			 : "=r" (val));
```

Comment:
minor: This assembly sequence is a bit ugly. Could you define an assembly helper function somewhere and call it from here?